server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/MailboxFactory.scala

```diff
@@ -152,7 +163,60 @@ class MailboxFactory @Inject() (subscriptionManager: SubscriptionManager) {
             namespace = namespace,
             rights = rights,
             quotas = quotas,
-            isSubscribed = retrieveIsSubscribed)))
+            isSubscribed = isSubscribed)))
+    }
+  }
+
+  def create(id: MailboxId, mailboxSession: MailboxSession, quotaLoader: QuotaLoader): SMono[Either[Exception, Mailbox]] = {
+    try {
+      val messageManager: MessageManager = mailboxManager.getMailbox(id, mailboxSession)
+      val resolvedACL = messageManager.getResolvedAcl(mailboxSession)
+
+      val name: Either[IllegalArgumentException, MailboxName] = retrieveMailboxName(messageManager.getMailboxPath, mailboxSession)
+
+      val role: Option[Role] = getRole(messageManager.getMailboxPath, mailboxSession)
+      val sortOrder: SortOrder = getSortOrder(role)
+      val quotas: SMono[Quotas] = quotaLoader.getQuotas(messageManager.getMailboxPath)
+      val rights: Rights = getRights(resolvedACL)
+
+      val sanitizedCounters: MailboxCounters = messageManager.getMailboxCounters(mailboxSession).sanitize()
+      val unreadEmails: Either[NumberFormatException, UnsignedInt] = UnsignedInt.validate(sanitizedCounters.getUnseen)
+      val unreadThreads: Either[NumberFormatException, UnsignedInt] = UnsignedInt.validate(sanitizedCounters.getUnseen)
+      val totalEmails: Either[NumberFormatException, UnsignedInt] = UnsignedInt.validate(sanitizedCounters.getCount)
+      val totalThreads: Either[NumberFormatException, UnsignedInt] = UnsignedInt.validate(sanitizedCounters.getCount)
+
+      val namespace: MailboxNamespace = getNamespace(messageManager.getMailboxPath, mailboxSession)
+
+      val parentId: Option[MailboxId] = getParentPath(messageManager.getMailboxPath, mailboxSession)
+        .map(parentPath => mailboxManager.getMailbox(parentPath, mailboxSession))
+        .map(_.getId)
+
+      val myRights: MailboxRights = getMyRights(messageManager.getMailboxPath, resolvedACL, mailboxSession)
+
+      val isSubscribed: IsSubscribed = retrieveIsSubscribed(messageManager.getMailboxPath, mailboxSession)
+
+      MailboxValidation.validate(name, unreadEmails, unreadThreads, totalEmails, totalThreads) match {
+        case Left(error) => SMono.just(Left(error))
+        case scala.Right(mailboxValidation) => SMono.fromPublisher(quotas)
+          .map(quotas => scala.Right(
+            Mailbox(
+              id = id,
+              name = mailboxValidation.mailboxName,
+              parentId = parentId,
+              role = role,
+              sortOrder = sortOrder,
+              unreadEmails = mailboxValidation.unreadEmails,
+              totalEmails = mailboxValidation.totalEmails,
+              unreadThreads = mailboxValidation.unreadThreads,
+              totalThreads = mailboxValidation.totalThreads,
+              myRights = myRights,
+              namespace = namespace,
+              rights = rights,
+              quotas = quotas,
+              isSubscribed = isSubscribed)))
```

Comment:
for example on an error in  SMono.fromPublisher(quotas) what should we do ? return a Mono with a Left  ?

Reply:
It's not gonna raise an error in the level above in this case?

Reply:
Yes, but i pointed it out as an example of that the return type doesn't tel us at which level we should handle each type of exception.